File: src/tmux.ts
import { execSync, execFile as execFileCb } from "node:child_process";
import { promisify } from "node:util";

const execFileAsync = promisify(execFileCb);

export interface TmuxPane {
  sessionName: string;
  windowIndex: number;
  windowName: string;
  paneIndex: number;
  paneId: string; // e.g., "%0"
  target: string; // e.g., "session:0.0"
  active: boolean;
  cols: number;
  rows: number;
  // Geometry for cropping (0-indexed, in character units)
  left: number;
  top: number;
  // Process info
  pid: number;
  process: string; // The actual running command (not just shell name)
}

export interface TmuxSession {
  name: string;
  windows: TmuxWindow[];
  activity?: number;
  path?: string;
}

export interface TmuxWindow {
  index: number;
  name: string;
  panes: TmuxPane[];
}

/**
 * Extract clean command name from ps output
 * Handles: /path/to/cmd, "cmd args", and combinations
 */
function extractCmdName(psOutput: string): string {
  // First, get just the first word (in case args are included)
  const firstWord = psOutput.split(/\s+/)[0];
  // Then extract basename from path
  const lastSlash = firstWord.lastIndexOf("/");
  return lastSlash >= 0 ? firstWord.slice(lastSlash + 1) : firstWord;
}

/**
 * Get child PIDs of a process (portable across macOS and Linux)
 */
function getChildPids(ppid: number): number[] {
  try {
    // Use ps which works consistently across platforms
    const output = execSync(`ps -eo pid,ppid`, { encoding: "utf-8", stdio: ["pipe", "pipe", "pipe"] });
    const children: number[] = [];
    for (const line of output.split("\n")) {
      const parts = line.trim().split(/\s+/);
      if (parts.length >= 2 && parseInt(parts[1], 10) === ppid) {
        children.push(parseInt(parts[0], 10));
      }
    }
    return children;
  } catch {
    return [];
  }
}

/**
 * Get the effective process name for a pane
 * Traverses the process tree to find the actual running command (skipping shells and wrappers)
 */
function getEffectiveProcess(pid: number, currentCommand: string): string {
  // Commands to skip through when looking for the "real" process
  const wrappers = ["zsh", "bash", "sh", "fish", "tcsh", "csh", "-zsh", "-bash", "-sh", "npm", "npx", "node"];

  // If not a wrapper, return the current command
  if (!wrappers.includes(currentCommand)) {
    return currentCommand;
  }

  try {
    // Walk down the process tree until we find a non-wrapper or hit a leaf
    let currentPid = pid;
    let depth = 0;
    const maxDepth = 5; // Prevent infinite loops

    while (depth < maxDepth) {
      const children = getChildPids(currentPid);

      if (children.length === 0) {
        // No children - get this process's command
        if (currentPid !== pid) {
          const cmd = execSync(`ps -o comm= -p ${currentPid}`, { encoding: "utf-8", stdio: ["pipe", "pipe", "pipe"] }).trim();
          if (cmd) {
            return extractCmdName(cmd);
          }
        }
        return currentCommand;
      }

      // Check the first child
      const childPid = children[0];
      const childCmd = execSync(`ps -o comm= -p ${childPid}`, { encoding: "utf-8", stdio: ["pipe", "pipe", "pipe"] }).trim();

      if (!childCmd) {
        return currentCommand;
      }

      const cmdName = extractCmdName(childCmd);

      // If this is not a wrapper, we found our target
      if (!wrappers.includes(cmdName) && !wrappers.includes(`-${cmdName}`)) {
        return cmdName;
      }

      // Otherwise, continue down the tree
      currentPid = childPid;
      depth++;
    }

    return currentCommand;
  } catch {
    return currentCommand;
  }
}

/**
 * Check if tmux server is running
 */
export function isTmuxRunning(): boolean {
  try {
    execSync("tmux list-sessions", { stdio: "pipe" });
    return true;
  } catch {
    return false;
  }
}

/**
 * List all tmux sessions, windows, and panes
 */
export function listSessions(): TmuxSession[] {
  if (!isTmuxRunning()) {
    return [];
  }

  try {
    // Format: session_name:window_index:window_name:pane_index:pane_id:pane_active:pane_width:pane_height:pane_left:pane_top:pane_pid:pane_current_command
    const output = execSync(
      'tmux list-panes -a -F "#{session_name}:#{window_index}:#{window_name}:#{pane_index}:#{pane_id}:#{pane_active}:#{pane_width}:#{pane_height}:#{pane_left}:#{pane_top}:#{pane_pid}:#{pane_current_command}"',
      { encoding: "utf-8" }
    );

    const sessions = new Map<string, TmuxSession>();

    for (const line of output.trim().split("\n")) {
      if (!line) continue;

      const [sessionName, windowIndexStr, windowName, paneIndexStr, paneId, paneActiveStr, colsStr, rowsStr, leftStr, topStr, pidStr, currentCommand] = line.split(":");
      const windowIndex = parseInt(windowIndexStr, 10);
      const paneIndex = parseInt(paneIndexStr, 10);
      const active = paneActiveStr === "1";
      const cols = parseInt(colsStr, 10);
      const rows = parseInt(rowsStr, 10);
      const left = parseInt(leftStr, 10);
      const top = parseInt(topStr, 10);
      const pid = parseInt(pidStr, 10);
      const process = getEffectiveProcess(pid, currentCommand);

      const target = `${sessionName}:${windowIndex}.${paneIndex}`;

      const pane: TmuxPane = {
        sessionName,
        windowIndex,
        windowName,
        paneIndex,
        paneId,
        target,
        active,
        cols,
        rows,
        left,
        top,
        pid,
        process,
      };

      if (!sessions.has(sessionName)) {
        sessions.set(sessionName, { name: sessionName, windows: [] });
      }

      const session = sessions.get(sessionName)!;
      let window = session.windows.find((w) => w.index === windowIndex);
      if (!window) {
        window = { index: windowIndex, name: windowName, panes: [] };
        session.windows.push(window);
      }
      window.panes.push(pane);
    }

    // Sort windows and panes
    for (const session of sessions.values()) {
      session.windows.sort((a, b) => a.index - b.index);
      for (const window of session.windows) {
        window.panes.sort((a, b) => a.paneIndex - b.paneIndex);
      }
    }

    return Array.from(sessions.values());
  } catch (err) {
    console.error("Failed to list tmux sessions:", err);
    return [];
  }
}

/**
 * Capture the current content of a pane
 * Returns the rendered screen content (with ANSI codes stripped by default)
 */
export function capturePane(target: string, options: { escape?: boolean; start?: number; end?: number } = {}): string {
  if (!isTmuxRunning()) {
    throw new Error("tmux is not running");
  }

  try {
    // -p: print to stdout
    // -e: include escape sequences (ANSI codes)
    // -t: target pane
    const args = ["capture-pane", "-p", "-t", target];

    if (options.escape) {
      args.push("-e"); // Include ANSI escape sequences
    }

    if (options.start !== undefined) {
      args.push("-S", options.start.toString());
    }

    if (options.end !== undefined) {
      args.push("-E", options.end.toString());
    }

    const output = execSync(`tmux ${args.join(" ")}`, { encoding: "utf-8" });
    return output;
  } catch (err) {
    throw new Error(`Failed to capture pane ${target}: ${err}`);
  }
}

/**
 * Capture pane with ANSI escape codes for xterm.js rendering
 */
export function capturePaneWithEscapes(target: string): string {
  return capturePane(target, { escape: true });
}

/**
 * Send keys to a tmux pane
 * @param target - pane target (e.g., "session:0.0")
 * @param keys - keys to send
 * @param literal - if true, use -l flag to send literal string (no special key handling)
 */
export function sendKeys(target: string, keys: string, literal: boolean = false): void {
  if (!isTmuxRunning()) {
    throw new Error("tmux is not running");
  }

  try {
    const args = ["send-keys", "-t", target];
    if (literal) {
      args.push("-l");
    }
    args.push(keys);

    execSync(`tmux ${args.map(a => a.includes(" ") ? `"${a}"` : a).join(" ")}`, { encoding: "utf-8" });
  } catch (err) {
    throw new Error(`Failed to send keys to pane ${target}: ${err}`);
  }
}

/**
 * Resize a tmux pane
 */
export function resizePane(target: string, cols: number, rows: number): void {
  if (!isTmuxRunning()) {
    throw new Error("tmux is not running");
  }

  try {
    // Note: tmux resize-pane adjusts relative to current size or sets exact with -x/-y
    execSync(`tmux resize-pane -t ${target} -x ${cols} -y ${rows}`, { encoding: "utf-8" });
  } catch (err) {
    // Resize can fail if it would make pane too small or window is zoomed
    console.warn(`Failed to resize pane ${target}: ${err}`);
  }
}

/**
 * Get the dimensions of a session's current window
 * Returns the full window size (width x height in characters)
 */
export function getSessionDimensions(sessionName: string): { width: number; height: number } | null {
  if (!isTmuxRunning()) {
    return null;
  }

  try {
    const output = execSync(
      `tmux display-message -t "${sessionName}" -p "#{window_width}:#{window_height}"`,
      { encoding: "utf-8" }
    ).trim();

    const [widthStr, heightStr] = output.split(":");
    const width = parseInt(widthStr, 10);
    const height = parseInt(heightStr, 10);

    if (isNaN(width) || isNaN(height)) {
      return null;
    }

    return { width, height };
  } catch (err) {
    console.error(`Failed to get session dimensions for ${sessionName}:`, err);
    return null;
  }
}

/**
 * Kill a tmux pane
 */
export function killPane(target: string): void {
  if (!isTmuxRunning()) {
    throw new Error("tmux is not running");
  }

  try {
    execSync(`tmux kill-pane -t ${target}`, { encoding: "utf-8" });
  } catch (err) {
    throw new Error(`Failed to kill pane ${target}: ${err}`);
  }
}

export function killSession(name: string): void {
  if (!isTmuxRunning()) {
    throw new Error("tmux is not running");
  }

  try {
    execSync(`tmux kill-session -t ${name}`, { encoding: "utf-8" });
  } catch (err) {
    throw new Error(`Failed to kill session ${name}: ${err}`);
  }
}

/**
 * Get detailed info about a specific pane
 */
export function getPaneInfo(target: string): TmuxPane | null {
  if (!isTmuxRunning()) {
    return null;
  }

  try {
    const output = execSync(
      `tmux display-message -t ${target} -p "#{session_name}:#{window_index}:#{window_name}:#{pane_index}:#{pane_id}:#{pane_active}:#{pane_width}:#{pane_height}:#{pane_left}:#{pane_top}:#{pane_pid}:#{pane_current_command}"`,
      { encoding: "utf-8" }
    ).trim();

    const [sessionName, windowIndexStr, windowName, paneIndexStr, paneId, paneActiveStr, colsStr, rowsStr, leftStr, topStr, pidStr, currentCommand] = output.split(":");
    const pid = parseInt(pidStr, 10);
    const process = getEffectiveProcess(pid, currentCommand);

    return {
      sessionName,
      windowIndex: parseInt(windowIndexStr, 10),
      windowName,
      paneIndex: parseInt(paneIndexStr, 10),
      paneId,
      target,
      active: paneActiveStr === "1",
      cols: parseInt(colsStr, 10),
      rows: parseInt(rowsStr, 10),
      left: parseInt(leftStr, 10),
      top: parseInt(topStr, 10),
      pid,
      process,
    };
  } catch (err) {
    console.error(`Failed to get pane info for ${target}:`, err);
    return null;
  }
}

/**
 * Create a new tmux session (async)
 */
export async function createSessionAsync(name: string, cwd: string): Promise<void> {
  await execFileAsync("tmux", ["new-session", "-d", "-s", name, "-c", cwd]);
}

// ─── Async variants (non-blocking, for polling paths) ───────────────────────

/**
 * Fetch the entire process table in a single `ps` call.
 * Returns a map of PID → { ppid, comm } for in-memory tree walking.
 */
async function getProcessTable(): Promise<Map<number, { ppid: number; comm: string }>> {
  try {
    const { stdout } = await execFileAsync("ps", ["-eo", "pid=,ppid=,comm="], { encoding: "utf-8" });
    const table = new Map<number, { ppid: number; comm: string }>();
    for (const line of stdout.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed) continue;
      const match = trimmed.match(/^(\d+)\s+(\d+)\s+(.+)$/);
      if (match) {
        table.set(parseInt(match[1], 10), {
          ppid: parseInt(match[2], 10),
          comm: match[3],
        });
      }
    }
    return table;
  } catch {
    return new Map();
  }
}

/**
 * In-memory process tree walk using a pre-fetched process table.
 * Replaces N×depth execSync("ps") calls with pure lookups.
 */
function getEffectiveProcessFromTable(
  pid: number,
  currentCommand: string,
  table: Map<number, { ppid: number; comm: string }>,
): string {
  const wrappers = ["zsh", "bash", "sh", "fish", "tcsh", "csh", "-zsh", "-bash", "-sh", "npm", "npx", "node"];

  if (!wrappers.includes(currentCommand)) {
    return currentCommand;
  }

  let currentPid = pid;
  let depth = 0;
  const maxDepth = 5;

  while (depth < maxDepth) {
    const children: number[] = [];
    for (const [childPid, info] of table) {
      if (info.ppid === currentPid) {
        children.push(childPid);
      }
    }

    if (children.length === 0) {
      if (currentPid !== pid) {
        const proc = table.get(currentPid);
        if (proc) {
          const cmd = extractCmdName(proc.comm);
          if (cmd) return cmd;
        }
      }
      return currentCommand;
    }

    const childPid = children[0];
    const childInfo = table.get(childPid);
    if (!childInfo) return currentCommand;

    const cmdName = extractCmdName(childInfo.comm);
    if (!wrappers.includes(cmdName) && !wrappers.includes(`-${cmdName}`)) {
      return cmdName;
    }

    currentPid = childPid;
    depth++;
  }

  return currentCommand;
}

/**
 * Async version of listSessions.
 * Runs tmux list-panes and ps in parallel, then does in-memory tree walking.
 */
export async function listSessionsAsync(): Promise<TmuxSession[]> {
  try {
    const [tmuxOutput, processTable] = await Promise.all([
      execFileAsync(
        "tmux",
        ["list-panes", "-a", "-F",
          "#{session_name}:#{window_index}:#{window_name}:#{pane_index}:#{pane_id}:#{pane_active}:#{pane_width}:#{pane_height}:#{pane_left}:#{pane_top}:#{pane_pid}:#{pane_current_command}:#{session_activity}:#{session_path}"],
        { encoding: "utf-8" },
      ),
      getProcessTable(),
    ]);

    const sessions = new Map<string, TmuxSession>();

    for (const line of tmuxOutput.stdout.trim().split("\n")) {
      if (!line) continue;

      const parts = line.split(":");
      const sessionName = parts[0];
      const windowIndexStr = parts[1];
      const windowName = parts[2];
      const paneIndexStr = parts[3];
      const paneId = parts[4];
      const paneActiveStr = parts[5];
      const colsStr = parts[6];
      const rowsStr = parts[7];
      const leftStr = parts[8];
      const topStr = parts[9];
      const pidStr = parts[10];
      const currentCommand = parts[11];
      const sessionActivityStr = parts[12];
      // session_path may contain colons, so rejoin everything after field 13
      const sessionPath = parts.slice(13).join(":");

      const windowIndex = parseInt(windowIndexStr, 10);
      const paneIndex = parseInt(paneIndexStr, 10);
      const active = paneActiveStr === "1";
      const cols = parseInt(colsStr, 10);
      const rows = parseInt(rowsStr, 10);
      const left = parseInt(leftStr, 10);
      const top = parseInt(topStr, 10);
      const pid = parseInt(pidStr, 10);
      const process = getEffectiveProcessFromTable(pid, currentCommand, processTable);
      const sessionActivity = parseInt(sessionActivityStr, 10);

      const target = `${sessionName}:${windowIndex}.${paneIndex}`;

      const pane: TmuxPane = {
        sessionName, windowIndex, windowName, paneIndex, paneId, target,
        active, cols, rows, left, top, pid, process,
      };

      if (!sessions.has(sessionName)) {
        sessions.set(sessionName, { name: sessionName, windows: [], activity: sessionActivity, path: sessionPath || undefined });
      }

      const session = sessions.get(sessionName)!;
      let window = session.windows.find((w) => w.index === windowIndex);
      if (!window) {
        window = { index: windowIndex, name: windowName, panes: [] };
        session.windows.push(window);
      }
      window.panes.push(pane);
    }

    for (const session of sessions.values()) {
      session.windows.sort((a, b) => a.index - b.index);
      for (const window of session.windows) {
        window.panes.sort((a, b) => a.paneIndex - b.paneIndex);
      }
    }

    return Array.from(sessions.values());
  } catch {
    return [];
  }
}

/**
 * Async version of capturePane.
 */
export async function capturePaneAsync(target: string, options: { escape?: boolean; start?: number; end?: number } = {}): Promise<string> {
  const args = ["capture-pane", "-p", "-t", target];
  if (options.escape) args.push("-e");
  if (options.start !== undefined) args.push("-S", options.start.toString());
  if (options.end !== undefined) args.push("-E", options.end.toString());

  const { stdout } = await execFileAsync("tmux", args, { encoding: "utf-8" });
  return stdout;
}

/**
 * Async version of getSessionDimensions.
 */
export async function getSessionDimensionsAsync(sessionName: string): Promise<{ width: number; height: number } | null> {
  try {
    const { stdout } = await execFileAsync(
      "tmux",
      ["display-message", "-t", sessionName, "-p", "#{window_width}:#{window_height}"],
      { encoding: "utf-8" },
    );
    const [widthStr, heightStr] = stdout.trim().split(":");
    const width = parseInt(widthStr, 10);
    const height = parseInt(heightStr, 10);
    if (isNaN(width) || isNaN(height)) return null;
    return { width, height };
  } catch {
    return null;
  }
}

File: src-tauri/src/tmux.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use tokio::process::Command;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TmuxPane {
    pub session_name: String,
    pub window_index: u32,
    pub window_name: String,
    pub pane_index: u32,
    pub pane_id: String,
    pub target: String,
    pub active: bool,
    pub cols: u32,
    pub rows: u32,
    pub left: u32,
    pub top: u32,
    pub pid: u32,
    pub process: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub claude_session: Option<super::claude_sessions::ClaudeSession>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TmuxWindow {
    pub index: u32,
    pub name: String,
    pub panes: Vec<TmuxPane>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TmuxSession {
    pub name: String,
    pub windows: Vec<TmuxWindow>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub dimensions: Option<SessionDimensions>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub activity: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionDimensions {
    pub width: u32,
    pub height: u32,
}

/// Extract clean command name from ps output (basename of path)
fn extract_cmd_name(ps_output: &str) -> &str {
    let first_word = ps_output.split_whitespace().next().unwrap_or(ps_output);
    match first_word.rfind('/') {
        Some(pos) => &first_word[pos + 1..],
        None => first_word,
    }
}

/// Shell/wrapper commands to skip when walking the process tree
const WRAPPERS: &[&str] = &[
    "zsh", "bash", "sh", "fish", "tcsh", "csh", "-zsh", "-bash", "-sh", "npm", "npx", "node",
];

/// Fetch the entire process table in a single `ps` call.
async fn get_process_table() -> HashMap<u32, (u32, String)> {
    let output = Command::new("ps")
        .args(["-eo", "pid=,ppid=,comm="])
        .output()
        .await;

    let mut table = HashMap::new();
    if let Ok(output) = output {
        let stdout = String::from_utf8_lossy(&output.stdout);
        for line in stdout.lines() {
            let trimmed = line.trim();
            if trimmed.is_empty() {
                continue;
            }
            // Parse: PID PPID COMMAND
            let parts: Vec<&str> = trimmed.splitn(3, char::is_whitespace).collect();
            if parts.len() >= 3 {
                if let (Ok(pid), Ok(ppid)) = (parts[0].trim().parse::<u32>(), parts[1].trim().parse::<u32>()) {
                    table.insert(pid, (ppid, parts[2].trim().to_string()));
                }
            }
        }
    }
    table
}

/// Walk the process tree to find the real command (skip shells and wrappers)
fn get_effective_process_from_table(
    pid: u32,
    current_command: &str,
    table: &HashMap<u32, (u32, String)>,
) -> String {
    if !WRAPPERS.contains(&current_command) {
        return current_command.to_string();
    }

    let mut current_pid = pid;
    for _ in 0..5 {
        // Find children of current_pid
        let children: Vec<u32> = table
            .iter()
            .filter(|(_, (ppid, _))| *ppid == current_pid)
            .map(|(child_pid, _)| *child_pid)
            .collect();

        if children.is_empty() {
            if current_pid != pid {
                if let Some((_, comm)) = table.get(&current_pid) {
                    let cmd = extract_cmd_name(comm);
                    if !cmd.is_empty() {
                        return cmd.to_string();
                    }
                }
            }
            return current_command.to_string();
        }

        let child_pid = children[0];
        let child_info = match table.get(&child_pid) {
            Some(info) => info,
            None => return current_command.to_string(),
        };

        let cmd_name = extract_cmd_name(&child_info.1);
        if !WRAPPERS.contains(&cmd_name) {
            let prefixed = format!("-{}", cmd_name);
            if !WRAPPERS.contains(&prefixed.as_str()) {
                return cmd_name.to_string();
            }
        }

        current_pid = child_pid;
    }

    current_command.to_string()
}

/// Check if tmux server is running
pub async fn is_tmux_running() -> bool {
    Command::new("tmux")
        .args(["list-sessions"])
        .output()
        .await
        .map(|o| o.status.success())
        .unwrap_or(false)
}

/// List all tmux sessions with full pane info (async, non-blocking)
pub async fn list_sessions() -> Vec<TmuxSession> {
    let format_str = "#{session_name}:#{window_index}:#{window_name}:#{pane_index}:#{pane_id}:#{pane_active}:#{pane_width}:#{pane_height}:#{pane_left}:#{pane_top}:#{pane_pid}:#{pane_current_command}:#{session_activity}:#{session_path}";

    let (tmux_result, process_table) = tokio::join!(
        Command::new("tmux")
            .args(["list-panes", "-a", "-F", format_str])
            .output(),
        get_process_table()
    );

    let tmux_output = match tmux_result {
        Ok(o) if o.status.success() => String::from_utf8_lossy(&o.stdout).to_string(),
        _ => return vec![],
    };

    let mut sessions: HashMap<String, TmuxSession> = HashMap::new();

    for line in tmux_output.lines() {
        if line.is_empty() {
            continue;
        }

        let parts: Vec<&str> = line.splitn(14, ':').collect();
        if parts.len() < 12 {
            continue;
        }

        let session_name = parts[0].to_string();
        let window_index: u32 = parts[1].parse().unwrap_or(0);
        let window_name = parts[2].to_string();
        let pane_index: u32 = parts[3].parse().unwrap_or(0);
        let pane_id = parts[4].to_string();
        let active = parts[5] == "1";
        let cols: u32 = parts[6].parse().unwrap_or(80);
        let rows: u32 = parts[7].parse().unwrap_or(24);
        let left: u32 = parts[8].parse().unwrap_or(0);
        let top: u32 = parts[9].parse().unwrap_or(0);
        let pid: u32 = parts[10].parse().unwrap_or(0);
        let current_command = parts[11];
        let session_activity: u64 = parts.get(12).and_then(|s| s.parse().ok()).unwrap_or(0);
        // session_path may contain colons, so rejoin everything after field 13
        let session_path = if parts.len() > 13 {
            Some(parts[13..].join(":"))
        } else {
            None
        };

        let process = get_effective_process_from_table(pid, current_command, &process_table);
        let target = format!("{}:{}.{}", session_name, window_index, pane_index);

        let pane = TmuxPane {
            session_name: session_name.clone(),
            window_index,
            window_name: window_name.clone(),
            pane_index,
            pane_id,
            target,
            active,
            cols,
            rows,
            left,
            top,
            pid,
            process,
            claude_session: None,
        };

        let session = sessions.entry(session_name.clone()).or_insert_with(|| TmuxSession {
            name: session_name,
            windows: vec![],
            dimensions: None,
            activity: if session_activity > 0 {
                Some(session_activity)
            } else {
                None
            },
            path: session_path.filter(|p| !p.is_empty()),
        });

        if let Some(window) = session.windows.iter_mut().find(|w| w.index == window_index) {
            window.panes.push(pane);
        } else {
            session.windows.push(TmuxWindow {
                index: window_index,
                name: window_name,
                panes: vec![pane],
            });
        }
    }

    // Sort sessions by name (stable order — HashMap iteration is non-deterministic)
    let mut result: Vec<TmuxSession> = sessions.into_values().collect();
    result.sort_by(|a, b| a.name.cmp(&b.name));
    for session in &mut result {
        session.windows.sort_by_key(|w| w.index);
        for window in &mut session.windows {
            window.panes.sort_by_key(|p| p.pane_index);
        }
    }

    result
}

/// Get dimensions of a session's current window
pub async fn get_session_dimensions(session_name: &str) -> Option<SessionDimensions> {
    let output = Command::new("tmux")
        .args([
            "display-message",
            "-t",
            session_name,
            "-p",
            "#{window_width}:#{window_height}",
        ])
        .output()
        .await
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let trimmed = stdout.trim();
    let mut parts = trimmed.splitn(2, ':');
    let width: u32 = parts.next()?.parse().ok()?;
    let height: u32 = parts.next()?.parse().ok()?;
    Some(SessionDimensions { width, height })
}

/// Create a new tmux session (idempotent)
pub async fn create_session(name: &str, cwd: &str) -> Result<(), String> {
    // Check if session already exists
    let check = Command::new("tmux")
        .args(["has-session", "-t", name])
        .output()
        .await;

    if let Ok(o) = check {
        if o.status.success() {
            return Ok(()); // Already exists
        }
    }

    let output = Command::new("tmux")
        .args(["new-session", "-d", "-s", name, "-c", cwd])
        .output()
        .await
        .map_err(|e| format!("Failed to create session: {}", e))?;

    if output.status.success() {
        Ok(())
    } else {
        Err(format!(
            "tmux new-session failed: {}",
            String::from_utf8_lossy(&output.stderr)
        ))
    }
}

/// Kill a tmux session
pub async fn kill_session(name: &str) -> Result<(), String> {
    let output = Command::new("tmux")
        .args(["kill-session", "-t", name])
        .output()
        .await
        .map_err(|e| format!("Failed to kill session: {}", e))?;

    if output.status.success() {
        Ok(())
    } else {
        Err(format!(
            "tmux kill-session failed: {}",
            String::from_utf8_lossy(&output.stderr)
        ))
    }
}

/// Kill a tmux pane
pub async fn kill_pane(target: &str) -> Result<(), String> {
    let output = Command::new("tmux")
        .args(["kill-pane", "-t", target])
        .output()
        .await
        .map_err(|e| format!("Failed to kill pane: {}", e))?;

    if output.status.success() {
        Ok(())
    } else {
        Err(format!(
            "tmux kill-pane failed: {}",
            String::from_utf8_lossy(&output.stderr)
        ))
    }
}

/// Send keys to a tmux pane (literal text + Enter)
pub async fn send_keys_literal(target: &str, text: &str) -> Result<(), String> {
    let output = Command::new("tmux")
        .args(["send-keys", "-t", target, "-l", text])
        .output()
        .await
        .map_err(|e| format!("Failed to send keys: {}", e))?;

    if !output.status.success() {
        return Err(format!(
            "tmux send-keys failed: {}",
            String::from_utf8_lossy(&output.stderr)
        ));
    }

    // Send Enter
    Command::new("tmux")
        .args(["send-keys", "-t", target, "Enter"])
        .output()
        .await
        .map_err(|e| format!("Failed to send Enter: {}", e))?;

    Ok(())
}

/// Send Ctrl+C to a tmux pane
pub async fn send_interrupt(target: &str) -> Result<(), String> {
    let output = Command::new("tmux")
        .args(["send-keys", "-t", target, "C-c"])
        .output()
        .await
        .map_err(|e| format!("Failed to send interrupt: {}", e))?;

    if output.status.success() {
        Ok(())
    } else {
        Err(format!(
            "tmux send-keys C-c failed: {}",
            String::from_utf8_lossy(&output.stderr)
        ))
    }
}

/// Get pane info for a specific target
pub async fn get_pane_info(target: &str) -> Option<TmuxPane> {
    let format_str = "#{session_name}:#{window_index}:#{window_name}:#{pane_index}:#{pane_id}:#{pane_active}:#{pane_width}:#{pane_height}:#{pane_left}:#{pane_top}:#{pane_pid}:#{pane_current_command}";

    let output = Command::new("tmux")
        .args(["display-message", "-t", target, "-p", format_str])
        .output()
        .await
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let line = stdout.trim();
    let parts: Vec<&str> = line.splitn(12, ':').collect();
    if parts.len() < 12 {
        return None;
    }

    let pid: u32 = parts[10].parse().unwrap_or(0);
    // For single pane lookup, do a quick process table fetch
    let table = get_process_table().await;
    let process = get_effective_process_from_table(pid, parts[11], &table);

    Some(TmuxPane {
        session_name: parts[0].to_string(),
        window_index: parts[1].parse().unwrap_or(0),
        window_name: parts[2].to_string(),
        pane_index: parts[3].parse().unwrap_or(0),
        pane_id: parts[4].to_string(),
        target: target.to_string(),
        active: parts[5] == "1",
        cols: parts[6].parse().unwrap_or(80),
        rows: parts[7].parse().unwrap_or(24),
        left: parts[8].parse().unwrap_or(0),
        top: parts[9].parse().unwrap_or(0),
        pid,
        process,
        claude_session: None,
    })
}

/// Get pane's current working directory
pub async fn get_pane_cwd(target: &str) -> Option<String> {
    let output = Command::new("tmux")
        .args([
            "display-message",
            "-t",
            target,
            "-p",
            "#{pane_current_path}",
        ])
        .output()
        .await
        .ok()?;

    if !output.status.success() {
        return None;
    }

    let cwd = String::from_utf8_lossy(&output.stdout).trim().to_string();
    if cwd.is_empty() {
        None
    } else {
        Some(cwd)
    }
}

/// Capture last N lines of a pane with escape sequences
pub async fn capture_pane_with_escapes(target: &str, start_line: i32) -> Option<String> {
    let output = Command::new("tmux")
        .args([
            "capture-pane",
            "-t",
            target,
            "-p",
            "-e",
            "-S",
            &start_line.to_string(),
        ])
        .output()
        .await
        .ok()?;

    if output.status.success() {
        Some(String::from_utf8_lossy(&output.stdout).to_string())
    } else {
        None
    }
}

/// Check if a pane is showing Claude's orange thinking indicator
pub async fn is_pane_processing(target: &str) -> bool {
    let output = match capture_pane_with_escapes(target, -10).await {
        Some(o) => o,
        None => return false,
    };

    // Orange/salmon color range used by Claude Code thinking status
    // Pattern: \x1b[38;2;R;G;Bm where R=200-239, G=100-159, B=80-129
    let thinking_re = regex::Regex::new(
        r"\x1b\[38;2;(2[0-3][0-9]);(1[0-5][0-9]);([89][0-9]|1[0-2][0-9])m",
    )
    .unwrap();

    thinking_re.is_match(&output) && output.contains('\u{2026}') // ellipsis "…"
}
